### PR TITLE
feat: Type check DML

### PR DESCRIPTION
**Pull request by joshua-spacetime**
# Description of Changes

Added type checking and name resolution for SQL DML statements

# Expected complexity level and risk

2

# Testing

### CODE REVIEW

File: crates/planner/src/logical/bind.rs [hunk 1]
```diff
@@ -25,10 +25,121 @@ pub trait SchemaView {
...
+                let ty = Type::Var(types.into_boxed_slice());
+                let id = ctx.add(ty);
+                Ok((RelExpr::RelVar(schema, id), vec![(var.name, id)].into()))
+            }
+            ast::RelExpr::Ast(ast) => Ok((Self::type_ast(ctx, *ast, tx)?, Vars::default())),
+        }
+    }
+}
+
+struct SubChecker;
```
Comment (mamcx) [inline crates/planner/src/logical/bind.rs:94]:
```suggestion
/// Type checker for subscriptions
struct SubChecker;
```

File: crates/planner/src/logical/errors.rs [hunk 2]
```diff
@@ -89,6 +89,14 @@ pub enum Unsupported {
...
 }
 
+#[derive(Error, Debug)]
+#[error("Inserting a row with {values} values into {table} which has {fields} fields")]
+pub struct InsertError {
+    pub table: String,
+    pub values: usize,
+    pub fields: usize,
+}
+
```
Comment (mamcx) [inline crates/planner/src/logical/errors.rs:99]: Will be nice to know which `fields`/`values` failed here.

File: crates/planner/src/logical/stmt.rs [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+        }
+    }
+    let values = values.into_boxed_slice();
+    let filter = filter
+        .map(|expr| type_expr(ctx, &vars, expr, Some(TyId::BOOL)))
+        .transpose()?;
+    Ok(TableUpdate { schema, values, filter })
+}
+
+struct SqlChecker;
```
Comment (mamcx) [inline crates/planner/src/logical/stmt.rs:173]:
```suggestion
/// Type-checker for regular `SQL` queries
struct SqlChecker;
```

File: crates/planner/src/logical/errors.rs (hunk 2 above)
Comment (joshua-spacetime) [reply crates/planner/src/logical/errors.rs:107]: I added a TODO to that effect

File: crates/planner/src/logical/errors.rs [hunk 4]
```diff
@@ -89,6 +89,14 @@ pub enum Unsupported {
     UnqualifiedProjectExpr,
 }
 
+#[derive(Error, Debug)]
+#[error("Inserting a row with {values} values into {table} which has {fields} fields")]
+pub struct InsertError {
```
[the remark this answers is not in the document]
Comment (joshua-spacetime) [reply crates/planner/src/logical/errors.rs:102]: 👍 I updated the all the other error messages as well.

File: crates/planner/src/logical/ty.rs [hunk 5]
```diff
@@ -16,14 +16,20 @@ pub struct TyId(u32);
 
 impl TyId {
     /// The number of primitive types whose [TyId]s are statically defined.
-    pub const N: usize = 16;
+    pub const N: usize = 18;
```
Comment (mamcx) [inline crates/planner/src/logical/ty.rs:20]:
```suggestion
// The `ids here must match the `TyCtx::default` `vec` below

```

Comment (mamcx) [reply crates/planner/src/logical/ty.rs:20]: Also, add the note there...

**Review by mamcx — APPROVED** [review]

**Pull request merged**